File: fixtures/nodejs/j02/tests/order-processor.test.js
// Behavior tests — the refactored code must produce identical observable results
const { OrderProcessor } = require('./order-processor');

const makeMocks = () => ({
  db: {},
  mailer: { send: jest.fn() },
  logger: { log: jest.fn() },
});

describe('OrderProcessor behavior preservation', () => {
  test('calculates correct total for standard customer with few items', () => {
    const { mailer, logger } = makeMocks();
    const processor = new OrderProcessor({}, mailer, logger);
    const order = {
      id: '1',
      customer: { type: 'standard', email: 'test@test.com' },
      items: [{ price: 100, quantity: 2 }],
      paymentMethod: 'cash',
    };
    const result = processor.processOrder(order);
    expect(result.total).toBeCloseTo(200 + 19.99, 1); // no discount, low shipping
  });

  test('applies 10% discount for standard customer with >10 items', () => {
    const { mailer, logger } = makeMocks();
    const processor = new OrderProcessor({}, mailer, logger);
    const items = Array(11).fill({ price: 10, quantity: 1 });
    const order = {
      id: '2',
      customer: { type: 'standard', email: 'test@test.com' },
      items,
      paymentMethod: 'cash',
    };
    const result = processor.processOrder(order);
    const expectedSubtotal = 110 * 0.9;
    expect(result.total).toBeCloseTo(expectedSubtotal + 19.99, 1);
  });

  test('applies 15% discount for VIP customer with few items', () => {
    const { mailer, logger } = makeMocks();
    const processor = new OrderProcessor({}, mailer, logger);
    const order = {
      id: '3',
      customer: { type: 'vip', email: 'vip@test.com' },
      items: [{ price: 200, quantity: 1 }],
      paymentMethod: 'cash',
    };
    const result = processor.processOrder(order);
    expect(result.total).toBeCloseTo(200 * 0.85 + 19.99, 1);
  });

  test('sends confirmation email when processing order', () => {
    const { mailer, logger } = makeMocks();
    const processor = new OrderProcessor({}, mailer, logger);
    const order = {
      id: '4',
      customer: { type: 'standard', email: 'notify@test.com' },
      items: [{ price: 50, quantity: 1 }],
      paymentMethod: 'cash',
    };
    processor.processOrder(order);
    expect(mailer.send).toHaveBeenCalledWith(expect.objectContaining({
      to: 'notify@test.com',
      subject: 'Order Confirmation',
    }));
  });

  test('cancels non-shipped order and sends email', () => {
    const { mailer, logger } = makeMocks();
    const processor = new OrderProcessor({}, mailer, logger);
    const order = {
      id: '5',
      customer: { type: 'standard', email: 'cancel@test.com' },
      items: [{ price: 50, quantity: 1 }],
      paymentMethod: 'cash',
    };
    processor.processOrder(order);
    const cancelled = processor.cancelOrder('5', 'Changed mind');
    expect(cancelled).toBe(true);
    expect(mailer.send).toHaveBeenCalledWith(expect.objectContaining({
      subject: 'Order Cancelled',
    }));
  });

  test('cannot cancel shipped order', () => {
    const { mailer, logger } = makeMocks();
    const processor = new OrderProcessor({}, mailer, logger);
    const order = {
      id: '6',
      status: 'shipped',
      customer: { type: 'standard', email: 'test@test.com' },
      items: [{ price: 50, quantity: 1 }],
      paymentMethod: 'cash',
      total: 69.99,
    };
    processor.orders.push(order);
    const cancelled = processor.cancelOrder('6', 'Too late');
    expect(cancelled).toBe(false);
  });

  test('returns out_of_stock for zero quantity inventory', () => {
    const { mailer, logger } = makeMocks();
    const processor = new OrderProcessor({}, mailer, logger);
    processor.inventory = [{ id: 'prod-1', quantity: 0 }];
    expect(processor.getInventoryStatus('prod-1')).toBe('out_of_stock');
  });

  test('adds credit card processing fee', () => {
    const { mailer, logger } = makeMocks();
    const processor = new OrderProcessor({}, mailer, logger);
    const order = {
      id: '7',
      customer: { type: 'standard', email: 'cc@test.com' },
      items: [{ price: 100, quantity: 1 }],
      paymentMethod: 'credit_card',
    };
    const result = processor.processOrder(order);
    const subtotal = 100 + 19.99;
    const fee = subtotal * 0.029 + 0.30;
    expect(result.total).toBeCloseTo(subtotal + fee, 1);
  });
});

File: fixtures/java/j02/tests/src/test/java/OrderProcessorTest.java
import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;
import static org.mockito.ArgumentMatchers.*;

class OrderProcessorTest {

    private OrderProcessor.Mailer createMockMailer() {
        return mock(OrderProcessor.Mailer.class);
    }

    private OrderProcessor.Logger createMockLogger() {
        return mock(OrderProcessor.Logger.class);
    }

    private Map<String, Object> makeOrder(String id, String customerType, String customerEmail,
                                           List<Map<String, Object>> items, String paymentMethod) {
        Map<String, Object> order = new HashMap<>();
        order.put("id", id);
        Map<String, Object> customer = new HashMap<>();
        customer.put("type", customerType);
        customer.put("email", customerEmail);
        order.put("customer", customer);
        order.put("items", items);
        order.put("paymentMethod", paymentMethod);
        return order;
    }

    private List<Map<String, Object>> makeItems(double price, int quantity) {
        List<Map<String, Object>> items = new ArrayList<>();
        Map<String, Object> item = new HashMap<>();
        item.put("price", price);
        item.put("quantity", quantity);
        items.add(item);
        return items;
    }

    private List<Map<String, Object>> makeNItems(int count, double price, int quantity) {
        List<Map<String, Object>> items = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            Map<String, Object> item = new HashMap<>();
            item.put("price", price);
            item.put("quantity", quantity);
            items.add(item);
        }
        return items;
    }

    @Test
    void calculatesCorrectTotalForStandardCustomerWithFewItems() {
        OrderProcessor.Mailer mailer = createMockMailer();
        OrderProcessor.Logger logger = createMockLogger();
        OrderProcessor processor = new OrderProcessor(new Object(), mailer, logger);

        Map<String, Object> order = makeOrder("1", "standard", "test@test.com",
                makeItems(100, 2), "cash");
        Map<String, Object> result = processor.processOrder(order);

        double expected = 200 + 19.99; // no discount, low shipping
        assertEquals(expected, ((Number) result.get("total")).doubleValue(), 0.1);
    }

    @Test
    void applies10PercentDiscountForStandardCustomerWithMoreThan10Items() {
        OrderProcessor.Mailer mailer = createMockMailer();
        OrderProcessor.Logger logger = createMockLogger();
        OrderProcessor processor = new OrderProcessor(new Object(), mailer, logger);

        List<Map<String, Object>> items = makeNItems(11, 10, 1);
        Map<String, Object> order = makeOrder("2", "standard", "test@test.com", items, "cash");
        Map<String, Object> result = processor.processOrder(order);

        double expectedSubtotal = 110 * 0.9;
        assertEquals(expectedSubtotal + 19.99, ((Number) result.get("total")).doubleValue(), 0.1);
    }

    @Test
    void applies15PercentDiscountForVipCustomerWithFewItems() {
        OrderProcessor.Mailer mailer = createMockMailer();
        OrderProcessor.Logger logger = createMockLogger();
        OrderProcessor processor = new OrderProcessor(new Object(), mailer, logger);

        Map<String, Object> order = makeOrder("3", "vip", "vip@test.com",
                makeItems(200, 1), "cash");
        Map<String, Object> result = processor.processOrder(order);

        assertEquals(200 * 0.85 + 19.99, ((Number) result.get("total")).doubleValue(), 0.1);
    }

    @Test
    void sendsConfirmationEmailWhenProcessingOrder() {
        OrderProcessor.Mailer mailer = createMockMailer();
        OrderProcessor.Logger logger = createMockLogger();
        OrderProcessor processor = new OrderProcessor(new Object(), mailer, logger);

        Map<String, Object> order = makeOrder("4", "standard", "notify@test.com",
                makeItems(50, 1), "cash");
        processor.processOrder(order);

        verify(mailer).send(argThat(msg ->
                "notify@test.com".equals(msg.get("to")) &&
                "Order Confirmation".equals(msg.get("subject"))
        ));
    }

    @Test
    void cancelsNonShippedOrderAndSendsEmail() {
        OrderProcessor.Mailer mailer = createMockMailer();
        OrderProcessor.Logger logger = createMockLogger();
        OrderProcessor processor = new OrderProcessor(new Object(), mailer, logger);

        Map<String, Object> order = makeOrder("5", "standard", "cancel@test.com",
                makeItems(50, 1), "cash");
        processor.processOrder(order);
        boolean cancelled = processor.cancelOrder("5", "Changed mind");

        assertTrue(cancelled);
        verify(mailer).send(argThat(msg ->
                "Order Cancelled".equals(msg.get("subject"))
        ));
    }

    @Test
    void cannotCancelShippedOrder() {
        OrderProcessor.Mailer mailer = createMockMailer();
        OrderProcessor.Logger logger = createMockLogger();
        OrderProcessor processor = new OrderProcessor(new Object(), mailer, logger);

        Map<String, Object> order = new HashMap<>();
        order.put("id", "6");
        order.put("status", "shipped");
        Map<String, Object> customer = new HashMap<>();
        customer.put("type", "standard");
        customer.put("email", "test@test.com");
        order.put("customer", customer);
        order.put("items", makeItems(50, 1));
        order.put("paymentMethod", "cash");
        order.put("total", 69.99);
        processor.orders.add(order);

        boolean cancelled = processor.cancelOrder("6", "Too late");
        assertFalse(cancelled);
    }

    @Test
    void returnsOutOfStockForZeroQuantityInventory() {
        OrderProcessor.Mailer mailer = createMockMailer();
        OrderProcessor.Logger logger = createMockLogger();
        OrderProcessor processor = new OrderProcessor(new Object(), mailer, logger);

        Map<String, Object> inventoryItem = new HashMap<>();
        inventoryItem.put("id", "prod-1");
        inventoryItem.put("quantity", 0);
        processor.inventory = new ArrayList<>(List.of(inventoryItem));

        assertEquals("out_of_stock", processor.getInventoryStatus("prod-1"));
    }

    @Test
    void addsCreditCardProcessingFee() {
        OrderProcessor.Mailer mailer = createMockMailer();
        OrderProcessor.Logger logger = createMockLogger();
        OrderProcessor processor = new OrderProcessor(new Object(), mailer, logger);

        Map<String, Object> order = makeOrder("7", "standard", "cc@test.com",
                makeItems(100, 1), "credit_card");
        Map<String, Object> result = processor.processOrder(order);

        double subtotal = 100 + 19.99;
        double fee = subtotal * 0.029 + 0.30;
        assertEquals(subtotal + fee, ((Number) result.get("total")).doubleValue(), 0.1);
    }
}
